gef.py

```diff
@@ -8603,6 +8616,118 @@ def get_settings_path(self):
+            raise gdb.GdbError("No {} section".format(self._section_ or self._zone_))
+
+        addr = long(offset) if offset.address is None else long(offset.address)
+        return long(base_address + addr)
+
+    def get_base_address(self):
+        if self._section_:
+            section = process_lookup_path(self._section_)
+            if section:
+                return section.page_start
+        elif self._zone_:
+            zone = file_lookup_name_path(self._zone_, get_filepath())
+            if zone:
+                return zone.zone_start
+        return None
+
+@register_function
+class StackOffsetFunction(GenericOffsetFunction):
+    """Return the current stack base address plus an optional offset."""
+    _function_ = "_stack"
+    _section_ = "[stack]"
+    _zone_ = None
+
+@register_function
+class HeapBaseFunction(GenericOffsetFunction):
+    """Return the current heap base address plus an optional offset."""
+    _function_ = "_heap"
+    _section_ = "[heap]"
+    _zone_ = None
+
+@register_function
+class PieBaseFunction(GenericOffsetFunction):
+    """Return the current pie base address plus an optional offset."""
+    _function_ = "_pie"
+    _zone_ = None
+    @property
+    def _section_(self):
+        return get_filepath()
+
+@register_function
+class BssBaseFunction(GenericOffsetFunction):
+    """Return the current bss base address plus the given offset."""
+    _function_ = "_bss"
+    _zone_ = ".bss"
+    _section_ = None
+
+@register_function
+class GotBaseFunction(GenericOffsetFunction):
+    """Return the current bss base address plus the given offset."""
+    _function_ = "_got"
+    _zone_ = ".got"
+    _section_ = None
+
+@register_command
+class GefFunctionsCommand(GenericCommand):
+    """List the convenience functions provided by GEF."""
+    _cmdline_ = "functions"
+    _syntax_ = _cmdline_
+
+    def __init__(self):
+        super(GefFunctionsCommand, self).__init__()
+        self.docs = []
+        self.setup()
+        return
+
+    def setup(self):
+        global __gef__
+        for function in __gef__.loaded_functions:
+            self.add_function_to_doc(function)
+        self.__doc__ = "\n".join(sorted(self.docs))
+
+    def add_function_to_doc(self, function):
+        """Add function to documentation."""
+        doc = getattr(function, "__doc__", "").lstrip()
+        doc = "\n                         ".join(doc.split("\n"))
+        syntax = getattr(function, "_syntax_", "").lstrip()
+        w = max(1, get_terminal_size()[1] - 29)  # use max() to avoid zero or negative numbers
+        msg = "{syntax:<25s} -- {help:{w}s}".format(syntax=syntax, help=Color.greenify(doc), w=w)
+        self.docs.append(msg)
+        return
+
+    def do_invoke(self, argv):
+        self.dont_repeat()
+        gef_print(titlify("GEF - Convenience Functions"))
+        gef_print("These functions can be used as arguments to other commands to dynamically calculate values, eg: {:s}\n"
```

Comment:
This is annoying but we keep lines under 120 chars (easy to tell since that's max width github shows)